hardware-wallet-tests/multiple-hardware-wallets.ts

```diff
@@ -0,0 +1,52 @@
+/* eslint-disable jest/no-standalone-expect */
+import expect from 'expect';
+
+import {
+  createAndRegisterHardwareWalletChannels,
+  createHardwareWalletConnectionChannel,
+  createSequentialResult,
+  initLedgerChannel,
+  createSequentialPromptMessages,
+} from './utils';
+
+export const run = () => {
+  expect.assertions(2);
+
+  createAndRegisterHardwareWalletChannels();
+
+  const hardwareWalletConnectionChannel = createHardwareWalletConnectionChannel();
+
+  const promptMessages = createSequentialPromptMessages([
+    'Start test runner',
+    'Plug Ledger Nano S to your computer',
+    'Plug Ledger Nano X to your computer',
+  ]);
+
+  promptMessages();
+
+  const expectedSequence = createSequentialResult([
+    {
+      disconnected: false,
+      deviceModel: 'nanoS',
+    },
+    {
+      disconnected: false,
+      deviceModel: 'nanoX',
+    },
+  ]);
+
+  return new Promise((resolve) => {
+    hardwareWalletConnectionChannel.onReceive(
+      async (params: { path: string; deviceModel: string }) => {
+        const [expectedValue, isOver] = expectedSequence();
+        expect(params).toEqual(expectedValue);
+
+        if (isOver) {
```

Comment:
Isn't this the expected outcome ?

```
- received additional unwanted messages received after expected set of messages
- received less messages than it is expected - in this case the test will stack in the dead point because the isOver variable will be false so the implementation will wait for additional message which may not arrive.
```

This test is assuming the current flow:
1- Nano S is connected
2- Nano X is connected

We could consider the code is buggy or requirements were changed in case the messages emitted were different than the one expected. 

As for the `setTimeout` we could use @marcin-mazurek proposed solution to use wait user to press enter before moving to next action (although this would require more effort)

Reply:
I think these cases mentioned above should not be considered as expected. We test the `hardwareWalletConnectionChannel` expecting specific messages in specific order. If there will be some additional messages, it may be incorrect behaviour. I think we don't want Daedalus to receive additional unwanted messages.

The second one is a bigger problem IMO, because the execution of this test script may be stuck forever, while the test should rise an error after some time. Imagine having such situation in the CI. We wouldn't want the pipeline to be running forever, but rather fail after some time and release the resources. The `setTimeout` in my code proposal is there to prevent such case.

Reply:
Also I suggested an alternative implementation to your `createSequentialResult` util.:
```
let receivedMessages = [];
const expectedMessages = createExpectedMessagesSequence([
  // ...
]);
hardwareWalletConnectionChannel.onReceive(message => {
  receivedMessages = receivedMessages.concat(message);
  // ...
});
expect(receivedMessages).toEqual(expectedMessages);
```
So the `createExpectedMessagesSequence` would just do the extending the expected messages with common set of properties. IMO this approach is basically simpler and enough.

What do you think about it?

Reply:
**Flat structure**
The assertion has to live inside the `onReceive` listener since the current flow is reacting to `initLedgerChannel` published events. 

Another abstraction layer would be needed in order to achieve the flat structure that you are proposing.  In this case we could invest more time to make the HWStore itself testable in the future.

**Messages**

The `createExpectedMessagesSequence`  is using a queue, so it should fail in case there's any message coming out of the expected order. 

```
EXPECTED  A , B
MESSAGES A, C  // should fail

EXPECTED  A, B
MESSAGES A       // will stay there until a new msg arrives. 
```

**setTimout**
As for the `setTimeout`, this script is only intended to run locally but we could have some high level timeout to kill the process. The thing is that this script requires manual interaction.